File: 트리DP/우수마을_1949.java
package 트리DP;

import java.io.*;
import java.util.*;

/*
7
1000 3000 4000 1000 2000 2000 7000
1 2
2 3
4 3
4 5
6 2
6 7
 */
public class 우수마을_1949 {
	static int N;
	static int[][] dp;
	static int NICE = 0;
	static int NOT_NICE = 1;
	static boolean[] visited;
	static int[] populations;
	static ArrayList<Integer>[] adj;

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;

		N = Integer.parseInt(br.readLine());

		dp = new int[N + 1][2];
		visited = new boolean[N + 1];
		populations = new int[N + 1];
		adj = new ArrayList[N + 1];

		for (int i = 0; i < N + 1; i++) {
			adj[i] = new ArrayList<>();
		}

		st = new StringTokenizer(br.readLine());
		for (int i = 1; i < N + 1; i++) {
			populations[i] = Integer.parseInt(st.nextToken());
		}

		for (int i = 0; i < N - 1; i++) {
			st = new StringTokenizer(br.readLine());
			int u = Integer.parseInt(st.nextToken());
			int v = Integer.parseInt(st.nextToken());

			adj[u].add(v);
			adj[v].add(u);
		}

		dfs(1);
		System.out.print(Math.max(dp[1][NICE], dp[1][NOT_NICE]));
	}

	private static void dfs(int root) {
		dp[root][NICE] += populations[root];
		visited[root] = true;

		for (int neighbors : adj[root]) {
			if (visited[neighbors]) continue;

			dfs(neighbors);

			dp[root][NOT_NICE] += Math.max(dp[neighbors][NICE], dp[neighbors][NOT_NICE]);
			dp[root][NICE] += dp[neighbors][NOT_NICE];
		}
	}
}

File: 코틀린_코테스터디/week8/우수마을_1949.kt
package 코틀린_코테스터디.week8

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.math.max


class 우수마을_1949 {
    private var N = 0
    private lateinit var dp: Array<Array<Int>>
    private var NICE = 0
    private var NOT_NICE = 1
    private lateinit var visited: Array<Boolean>
    private lateinit var populations: Array<Int>
    private lateinit var adj: Array<MutableList<Int>>

    fun solution() {
        val br = BufferedReader(InputStreamReader(System.`in`))
        var st: StringTokenizer

        N = br.readLine().toInt()

        dp = Array(N + 1) { Array(2) { 0 } }
        visited = Array(N + 1) { false }
        populations = Array(N + 1) { 0 }
        adj = Array(N + 1) { (mutableListOf()) }

        for (i in 0 until N + 1) {
            adj[i] = mutableListOf()
        }

        st = StringTokenizer(br.readLine())
        for (i in 1 until N + 1) {
            populations[i] = st.nextToken().toInt()
        }

        for (i in 0 until N - 1) {
            st = StringTokenizer(br.readLine())
            val u = st.nextToken().toInt()
            val v = st.nextToken().toInt()
            adj[u].add(v)
            adj[v].add(u)
        }

        dfs(1)
        println(max(dp[1][NICE], dp[1][NOT_NICE]))
    }

    private fun dfs(root: Int) {
        dp[root][NICE] += populations[root]
        visited[root] = true

        for (neighbors in adj[root]) {
            if (visited[neighbors]) continue

            dfs(neighbors)

            dp[root][NOT_NICE] += max(dp[neighbors][NICE], dp[neighbors][NOT_NICE])
            dp[root][NICE] += dp[neighbors][NOT_NICE]
        }
    }
}

fun main() {
    val problem = 우수마을_1949()
    problem.solution()
}